x-pack/plugins/observability_solution/infra/public/components/asset_details/tabs/dashboards/dashboards.tsx

```diff
@@ -0,0 +1,214 @@
+ */
+import React, { useCallback, useEffect, useState } from 'react';
+
+import { i18n } from '@kbn/i18n';
+import {
+  EuiPanel,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiTitle,
+  EuiSpacer,
+  EuiEmptyPrompt,
+  EuiLoadingLogo,
+} from '@elastic/eui';
+
+import { ViewMode } from '@kbn/embeddable-plugin/public';
+import {
+  AwaitingDashboardAPI,
+  DashboardCreationOptions,
+  DashboardRenderer,
+} from '@kbn/dashboard-plugin/public';
+
+import { buildPhraseFilter, Filter } from '@kbn/es-query';
+import type { DataView } from '@kbn/data-views-plugin/common';
+import { HOST_FIELD } from '../../../../../common/constants';
+import type {
+  DashboardIdItem,
+  DashboardItemWithTitle,
+  InfraCustomDashboardAssetType,
+} from '../../../../../common/custom_dashboards';
+
+import { EmptyDashboards } from './empty_dashboards';
+import { EditDashboard, GotoDashboard, LinkDashboard, UnlinkDashboard } from './actions';
+import { useCustomDashboard } from '../../hooks/use_custom_dashboards';
+import { useDatePickerContext } from '../../hooks/use_date_picker';
+import { useAssetDetailsRenderPropsContext } from '../../hooks/use_asset_details_render_props';
+import { FETCH_STATUS, useDashboardFetcher } from '../../hooks/use_dashboards_fetcher';
+import { useDataViewsContext } from '../../hooks/use_data_views';
+import { DashboardSelector } from './dashboard_selector';
+import { ContextMenu } from './context_menu';
+import { useAssetDetailsUrlState } from '../../hooks/use_asset_details_url_state';
+
+const fieldByAssetType = {
+  host: HOST_FIELD,
+} as Record<InfraCustomDashboardAssetType, string>;
+
+export function getFilterByAssetName(
+  assetName: string,
+  assetType: InfraCustomDashboardAssetType,
+  dataView: DataView
+): Filter[] {
+  const filters: Filter[] = [];
+
+  const assetNameField = dataView.getFieldByName(fieldByAssetType[assetType]);
+  if (assetNameField) {
+    const assetNameFilter = buildPhraseFilter(assetNameField, assetName, dataView);
+    filters.push(assetNameFilter);
+  }
+
+  return filters;
+}
+
+export function Dashboards() {
+  const { dateRange } = useDatePickerContext();
+  const { asset } = useAssetDetailsRenderPropsContext();
+  const [dashboard, setDashboard] = useState<AwaitingDashboardAPI>();
+  const [customDashboards, setCustomDashboards] = useState<DashboardItemWithTitle[]>([]);
+  const [currentDashboard, setCurrentDashboard] = useState<DashboardItemWithTitle>();
+  const { data: allAvailableDashboards, status } = useDashboardFetcher();
+  const { metrics } = useDataViewsContext();
+  const [urlState, setUrlState] = useAssetDetailsUrlState();
+
+  const { dashboards, loading, reload } = useCustomDashboard({ assetType: asset.type });
+
+  useEffect(() => {
+    const filteredCustomDashboards = (dashboards?.dashboardIdList ?? []).reduce(
+      (result: DashboardItemWithTitle[], customDashboard: DashboardIdItem) => {
+        const matchedDashboard = allAvailableDashboards.find(({ id }) => id === customDashboard.id);
```

Comment:
Not sure if this is relevant but you added an O(2ˆn) time complexity here. We could make a map of the allAvailableDashboards, and do:
```suggestion
        const matchedDashboard = allAvailableDashboardsMap[customDashboard.id];
```